libs/user-facing-errors/src/query_engine/validation.rs

```diff
@@ -1,15 +1,621 @@
+    ///
+    /// Say the schema type for user.id is `Int`
+    ///
+    /// The example json query will fail, as it's trying to pass a string instead.
+    ///
+    /// {
+    ///     "action": "findMany",
+    ///     "modelName": "User",
+    ///     "query": {
+    ///         "arguments": {
+    ///             "where": {
+    ///                 "dob": "invalid date"
+    ///             }
+    ///         },
+    ///         "selection": {
+    ///             "$scalars": true
+    ///         }
+    ///     }
+    /// }
+    pub fn invalid_argument_value(
+        selection_path: Vec<String>,
+        argument_path: Vec<String>,
+        value: String,
+        expected_argument_type: String,
+        underlying_err: Option<Box<dyn error::Error>>,
+    ) -> Self {
+        let argument_name = argument_path.last().expect("Argument path cannot not be empty");
+
+        let (message, meta) = if let Some(err) = underlying_err {
+            let err_msg = err.to_string();
+            let message = format!(
+                "Invalid argument agument value. `{}` is not a valid `{}`. Underlying error: {}",
+                value, expected_argument_type, &err_msg
+            );
+            let argument = ArgumentDescription::new(argument_name.to_owned(), vec![expected_argument_type]);
+            let meta = json!({"argumentPath": argument_path, "argument": argument, "underlying_error": &err_msg});
+            (message, Some(meta))
+        } else {
+            let message = format!(
+                "Invalid argument agument value. `{}` is not a valid `{}`",
+                value, &expected_argument_type
+            );
+            let argument = ArgumentDescription::new(argument_name.to_owned(), vec![expected_argument_type]);
+            let meta = json!({"argumentPath": argument_path, "argument": argument, "selectionPath": selection_path, "underlying_error": serde_json::Value::Null});
+            (message, Some(meta))
+        };
+
+        ValidationError {
+            kind: ValidationErrorKind::InvalidArgumentValue,
+            message,
+            meta,
+        }
+    }
+
+    /// Creates an [`ValidationErrorKind::SomeFieldsMissing`] kind of error, which happens when
+    /// there are some fields missing from a query
+    pub fn some_fields_missing(
+        selection_path: Vec<String>,
+        argument_path: Vec<String>,
+        min_field_count: Option<usize>,
+        max_field_count: Option<usize>,
+        required_fields: Option<Vec<String>>,
+        provided_field_count: usize,
+    ) -> Self {
+        let constraints =
+            InputTypeConstraints::new(min_field_count, max_field_count, required_fields, provided_field_count);
+
+        let message = format!("Some fields are missing: {}", constraints);
+        ValidationError {
+            kind: ValidationErrorKind::SomeFieldsMissing,
+            message,
+            meta: Some(json!({ "argumentPath": argument_path, "selectionPath": selection_path })),
```

Comment:
This is a good suggestion, I also thought about it. I think type-safety is a good property, OTOH, there is quite a variance in the meta attributes of different errors. Because we have tests asserting those formats, and the benefit of having such structs will be mainly lexica, I think it's fine to rely on the tests for that and not require the different metadata to have a different struct value.

Having said that this is a weak preference, and I'm fine changing it there is a hidden benefit I haven't considered.